### ISSUE TITLE
lexical warnings and inheritance

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#5770]
Migrated from [rt.perl.org#15395](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=15395) (status was 'resolved')

Searchable as RT15395$

**Comment by p5pRT** [comment]
### From anno4000@lublin.zrz.tu-berlin.de


This is a bug report for perl from anno4000@&#8203;lublin\.zrz\.tu\-berlin\.de\,
generated with the help of perlbug 1\.34 running under perl v5\.8\.0\.




---
The behavior of lexical warnings can be confusing if inheritance
\(or @&#8203;CARP\_NOT\) is involved\.  In particular\, @&#8203;INC \(and @&#8203;CARP\_NOT\)
influence where Carp \(as employed by warnings\) locates an error\.
However\, warnings has its own idea where an error is located\,
which doesn't look at @&#8203;INC and @&#8203;CARP\_NOT\.

This leads to the confusing behavior that a user may see a warning
in category "cat" from "file xxx line nnn"\, but putting \(say\)
"no warnings 'cat'" in line nnn \- 1 has no effect\, because
warnings takes its lexical scope from a deeper point in the call
stack\.  The following program shows this&#8203;:


&nbsp;    \#\!/usr/bin/perl
&nbsp;    use strict; $| = 1;
&nbsp;    
&nbsp;    BEGIN \{ 
&nbsp;        package Foo;
&nbsp;        use strict; \# @&#8203;^~\`
&nbsp;    
&nbsp;        our @&#8203;CARP\_NOT = qw\( Bar\);
&nbsp;    
&nbsp;        sub foofunc \{
&nbsp;            Bar&#8203;::barfunc\(\);
&nbsp;        \}
&nbsp;    \}
&nbsp;    
&nbsp;    BEGIN \{
&nbsp;    
&nbsp;        package Bar;
&nbsp;        use strict; \# @&#8203;^~\`
&nbsp;    
&nbsp;        use warnings&#8203;::register;
&nbsp;    
&nbsp;        sub barfunc \{
&nbsp;        
&nbsp;            warnings&#8203;::warn\( "from barfunc\, obligatorily"\);
&nbsp;            warnings&#8203;::warnif\( "from barfunc with love"\);
&nbsp;            print "barfunc done\\n";
&nbsp;        \}
&nbsp;    
&nbsp;    \}
&nbsp;    
&nbsp;    package main;  \# main program
&nbsp;    use warnings 'Bar';
&nbsp;    print "Bar warnings on&#8203;:\\n";
&nbsp;    Foo&#8203;::foofunc;
&nbsp;    print "=====================================\\n";
&nbsp;    
&nbsp;    no warnings 'Bar';
&nbsp;    print "Bar warnings off&#8203;:\\n";
&nbsp;    Foo&#8203;::foofunc;
&nbsp;    print "=====================================\\n";
&nbsp;    
&nbsp;    use warnings FATAL => 'Bar';
&nbsp;    print "Bar warnings FATAL&#8203;:\\n";
&nbsp;    Foo&#8203;::foofunc;
&nbsp;    print "=====================================\\n";
&nbsp;    
&nbsp;    \_\_END\_\_

The patch below corrects this by using Carp's notion of the error
location also in warnings\.  Affected files are warnings\.pl and
lib/Carp\.pm\, and\, for convenience\, lib/warnings\.pm\.

\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_

diff \-p0 \-rc orig/lib/Carp\.pm patched/lib/Carp\.pm
\*\*\* orig/lib/Carp\.pm    Tue Jul 23 12&#8203;:13&#8203;:27 2002
\-\-\- patched/lib/Carp\.pm    Tue Jul 23 12&#8203;:01&#8203;:52 2002
\*\*\*\*\*\*\*\*\*\*\*\*\*\*\* require Exporter;
\*\*\* 135 \*\*\*\*
\! @&#8203;EXPORT\_OK = qw\(cluck verbose longmess shortmess\);
\-\-\- 135 \-\-\-\-
\! @&#8203;EXPORT\_OK = qw\(cluck verbose longmess shortmess \_error\_loc\);
\*\*\*\*\*\*\*\*\*\*\*\*\*\*\* sub cluck   \{ warn longmess  @&#8203;\_ \}
\*\*\* 194 \*\*\*\*
\-\-\- 195\,202 \-\-\-\-
\+ 
\+ \# \_error\_loc is exported for warnings\, not general consumption\.  It helps
\+ \# warnings to have the same as Carp about where an error occured\.
\+ 
\+ sub \_error\_loc \{
\+     require Carp&#8203;::Heavy; \# no args\, no worries
\+     goto &short\_error\_loc; \# don't create another stack frame
\+ \}
diff \-p0 \-rc orig/lib/warnings\.pm patched/lib/warnings\.pm
\*\*\* orig/lib/warnings\.pm    Tue Jul 23 12&#8203;:13&#8203;:17 2002
\-\-\- patched/lib/warnings\.pm    Tue Jul 23 12&#8203;:30&#8203;:35 2002
\*\*\*\*\*\*\*\*\*\*\*\*\*\*\* See L\<perlmodlib/Pragmatic Modules> and 
\*\*\* 130 \*\*\*\*
\! use Carp ;
\-\-\- 130 \-\-\-\-
\! use Carp qw\(carp croak \_error\_loc\) ;
\*\*\*\*\*\*\*\*\*\*\*\*\*\*\* sub \_\_chk
\*\*\* 432\,436 \*\*\*\*
\!         for \($i = 2 ; $pkg = \(caller\($i\)\)\[0\] ; \+\+ $i\) \{
\!             last if $pkg ne $this\_pkg ;
\!         \}
\!         $i = 2
\!             if \!$pkg || $pkg eq $this\_pkg ;
\-\-\- 432 \-\-\-\-
\!         $i = \_error\_loc\(\); \# borrowed from Carp&#8203;::Heavy
diff \-p0 \-rc orig/warnings\.pl patched/warnings\.pl
\*\*\* orig/warnings\.pl    Tue Jul 23 12&#8203;:31&#8203;:34 2002
\-\-\- patched/warnings\.pl    Tue Jul 23 12&#8203;:29&#8203;:14 2002
\*\*\*\*\*\*\*\*\*\*\*\*\*\*\* See L\<perlmodlib/Pragmatic Modules> and 
\*\*\* 598 \*\*\*\*
\! use Carp ;
\-\-\- 598 \-\-\-\-
\! use Carp qw\(carp croak \_error\_loc\) ;
\*\*\*\*\*\*\*\*\*\*\*\*\*\*\* sub \_\_chk
\*\*\* 745\,749 \*\*\*\*
\!         for \($i = 2 ; $pkg = \(caller\($i\)\)\[0\] ; \+\+ $i\) \{
\!             last if $pkg ne $this\_pkg ;
\!         \}
\!         $i = 2
\!             if \!$pkg || $pkg eq $this\_pkg ;
\-\-\- 745 \-\-\-\-
\!         $i = \_error\_loc\(\); \# borrowed from Carp&#8203;::Heavy


---


---
Flags&#8203;:
&nbsp;    category=library
&nbsp;    severity=low


---
Site configuration information for perl v5\.8\.0&#8203;:

Configured by anno4000 at Mon Jul 22 11&#8203;:46&#8203;:29 MEST 2002\.

Summary of my perl5 \(revision 5\.0 version 8 subversion 0\) configuration&#8203;:
&nbsp;  Platform&#8203;:
&nbsp;    osname=linux\, osvers=2\.0\.36\, archname=i586\-linux
&nbsp;    uname='linux lublin\.zrz\.tu\-berlin\.de 2\.0\.36 \#2 mon jun 26 12&#8203;:24&#8203;:02 mest 2000 i586 unknown '
&nbsp;    config\_args='\-des'
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
&nbsp;    useperlio=define d\_sfio=undef uselargefiles=define usesocks=undef
&nbsp;    use64bitint=undef use64bitall=undef uselongdouble=undef
&nbsp;    usemymalloc=n\, bincompat5005=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='cc'\, ccflags ='\-I/usr/local/include \-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64'\,
&nbsp;    optimize='\-O2'\,
&nbsp;    cppflags='\-I/usr/local/include'
&nbsp;    ccversion=''\, gccversion='2\.7\.2\.3'\, gccosandvers=''
&nbsp;    intsize=4\, longsize=4\, ptrsize=4\, doublesize=8\, byteorder=1234
&nbsp;    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=12
&nbsp;    ivtype='long'\, ivsize=4\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\, lseeksize=4
&nbsp;    alignbytes=4\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='cc'\, ldflags =' \-L/usr/local/lib'
&nbsp;    libpth=/usr/local/lib /lib /usr/lib
&nbsp;    libs=\-lnsl \-lndbm \-lgdbm \-ldb \-ldl \-lm \-lc \-lposix \-lcrypt \-lutil
&nbsp;    perllibs=\-lnsl \-ldl \-lm \-lc \-lposix \-lcrypt \-lutil
&nbsp;    libc=/lib/libc\-2\.0\.7\.so\, so=so\, useshrplib=false\, libperl=libperl\.a
&nbsp;    gnulibc\_version='2\.0'
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags='\-rdynamic'
&nbsp;    cccdlflags='\-fpic'\, lddlflags='\-shared \-L/usr/local/lib'

Locally applied patches&#8203;:
&nbsp;    



---
@&#8203;INC for perl v5\.8\.0&#8203;:
&nbsp;    /opt/home/anno4000/lib/perl
&nbsp;    /usr/local/lib/perl5/5\.8\.0/i586\-linux
&nbsp;    /usr/local/lib/perl5/5\.8\.0
&nbsp;    /usr/local/lib/perl5/site\_perl/5\.8\.0/i586\-linux
&nbsp;    /usr/local/lib/perl5/site\_perl/5\.8\.0
&nbsp;    /usr/local/lib/perl5/site\_perl/5\.6\.1
&nbsp;    /usr/local/lib/perl5/site\_perl/5\.6\.0
&nbsp;    /usr/local/lib/perl5/site\_perl
&nbsp;    \.



---
Environment for perl v5\.8\.0&#8203;:
&nbsp;    HOME=/opt/home/anno4000
&nbsp;    LANG \(unset\)
&nbsp;    LANGUAGE \(unset\)
&nbsp;    LD\_LIBRARY\_PATH=/usr/X11R6/lib&#8203;:/usr/local/LessTif/Motif1\.2/lib
&nbsp;    LOGDIR \(unset\)
&nbsp;    PATH=/opt/home/anno4000/bin&#8203;:/sbin&#8203;:/usr/local/bin&#8203;:/bin&#8203;:/usr/bin&#8203;:/usr/sbin&#8203;:/usr/X11R6/bin&#8203;:/usr/openwin/bin&#8203;:/usr/games
&nbsp;    PERL5LIB=/opt/home/anno4000/lib/perl
&nbsp;    PERL\_BADLANG \(unset\)
&nbsp;    PERL\_PRIVLIB=/usr/local/lib/perl5/5\.8\.0
&nbsp;    PERL\_SITELIB=/usr/local/lib/perl5/site\_perl/5\.8\.0
&nbsp;    SHELL=/bin/tcsh

**Comment by p5pRT** [comment]
### From @hvds

Anno Siegel \(via RT\) \<perlbug@&#8203;perl\.org> wrote&#8203;:
:The behavior of lexical warnings can be confusing if inheritance
:\(or @&#8203;CARP\_NOT\) is involved\.  In particular\, @&#8203;INC \(and @&#8203;CARP\_NOT\)
:influence where Carp \(as employed by warnings\) locates an error\.
:However\, warnings has its own idea where an error is located\,
:which doesn't look at @&#8203;INC and @&#8203;CARP\_NOT\.

Thanks for the patch and associated analysis\.

I don't see any reason to export \_error\_loc \- why not just have
warnings\.pm call it as Carp&#8203;::error\_loc? I would strongly recommend
also including tests with any patch\, including specifically one or
more tests that would fail without the patch but succeed with it\.

If possible\, please also submit patches either as unified diffs
\(preferred\) or with context\.

Thanks\,

Hugo

**Comment by p5pRT** [comment]
### From siegel@zrz.tu-berlin.de

Hugo van der Sanden \(via RT\) wrote&#8203;:
> 
> Anno Siegel \(via RT\) \<perlbug@&#8203;perl\.org> wrote&#8203;:
> :The behavior of lexical warnings can be confusing if inheritance
> :\(or @&#8203;CARP\_NOT\) is involved\.  In particular\, @&#8203;INC \(and @&#8203;CARP\_NOT\)
> :influence where Carp \(as employed by warnings\) locates an error\.
> :However\, warnings has its own idea where an error is located\,
> :which doesn't look at @&#8203;INC and @&#8203;CARP\_NOT\.
> 
> Thanks for the patch and associated analysis\.
> 
> I don't see any reason to export \_error\_loc \- why not just have
> warnings\.pm call it as Carp&#8203;::error\_loc?

Sure\.  The point may be moot\, however\.

I had a little correspondence with Paul Marquess about warnings\.pm\,
and it turns out that using Carp for message delivery is perhaps not
the right way\.  The argument is that Carp is basically a DWIM module
which tries to guess the error location from \(mainly\) inheritance\.
Lexical warnings should give the user \(a module author\) full control
over where a warning is located on the call stack\.

I am working on a rewrite of warnings\.pm to that effect\.  Paul has
similar plans\, but says

&nbsp;    > I've more than enough on my plate at the moment\.  If you don't get
&nbsp;    > round to it\, I will eventually\.

So it's probably best to ignore my current patch for the moment\.

>                                         I would strongly recommend
> also including tests with any patch\, including specifically one or
> more tests that would fail without the patch but succeed with it\.

Yes\, that makes sense\.

> If possible\, please also submit patches either as unified diffs
> \(preferred\) or with context\.

Sorry if I didn't\.  That was an oversight\.

Regards\,

Anno

**Comment by p5pRT** [comment]
### From @cwest

Please consider re\-submitting this patch with the proposed updates by Hugo\.

**Comment by p5pRT** [comment]
### From siegel@zrz.tu-berlin.de

Casey West wrote&#8203;:

> Please consider re\-submitting this patch with the proposed updates by Hugo\.

Okay\, here is a revised patch\.

I have since seen that many standard modules take the equivalence of
Carp and warnings for granted\.  While I'm still unhappy with the situation
\(Carp and warnings growing together\)\, it seems too late for a change\.

The patch is against perl\-5\.8\.0\-RC3\.  It changes only warnings\.pl
and deposits a file bug\_15395\.t\.  This contains two tests that establish
that the patch has corrected the undesired behavior\.  The idea is to do

&nbsp;    cd perl\-5\.8\.0\-RC3
&nbsp;    patch \-p0 \< /path/to/patch
&nbsp;    perl bug\_15395\.t \# see it fail
&nbsp;    perl warnings\.pl \# update lib/warnings\.pm
&nbsp;    perl bug\_15395\.t \# see it pass

perl\-5\.8\.0\-RC3 need only contain warnings\.pl and lib/warnings\.pm\.

Anno
\-\- 

<details><summary>Inline Patch</summary>

```diff
--- ../../bleadperl/warnings.pl	Thu Jun 20 18:24:36 2002
+++ warnings.pl	Sat Aug 30 15:37:44 2003
@@ -742,16 +742,17 @@
 	$i -= 2 ;
     }
     else {
-        for ($i = 2 ; $pkg = (caller($i))[0] ; ++ $i) {
-            last if $pkg ne $this_pkg ;
-        }
-        $i = 2
-            if !$pkg || $pkg eq $this_pkg ;
+        $i = _error_loc(); # see where Carp will allocate the error
     }
 
     my $callers_bitmask = (caller($i))[9] ;
     return ($callers_bitmask, $offset, $i) ;
 }
+
+sub _error_loc {
+    require Carp::Heavy;
+    goto &Carp::short_error_loc; # don't introduce another stack frame
+}                                                             
 
 sub enabled
 {
--- /dev/null	Sat Aug 30 14:49:01 2003
+++ bug_15395.t	Sat Aug 30 15:46:56 2003
@@ -0,0 +1,56 @@
+#!perl
+
+use lib 'lib';
+
+use Test::More;
+
+my $n_tests;
+BEGIN { $n_tests += 2 }
+
+my ( $warn_cat, # warning category we'll try to control
+     $warn_msg, # the error message to catch
+);
+
+{ # lexical space for tests
+    no warnings; # keep stderr clean
+
+    package SomeModule;
+    use warnings::register;
+
+    BEGIN {
+        $warn_cat = __PACKAGE__;
+        $warn_msg = 'from ' . __PACKAGE__;
+    }
+
+    # a sub that generates a random warning
+    sub gen_warning {
+        warnings::warnif( $warn_msg);
+    }
+
+    package ClientModule;
+    # use SomeModule; (would go here)
+    our @CARP_NOT = ( $warn_cat); # deliver warnings to *our* client
+
+    # call_warner provokes a warning.  it is delivered to its caller,
+    # who should also be able to control it
+    sub call_warner {
+        SomeModule::gen_warning();
+    }
+
+# user
+
+package main;
+my $warn_line = __LINE__ + 3; # this line should be in the error message
+eval {
+    use warnings FATAL => $warn_cat; # we want to know if this works
+    ClientModule::call_warner();
+};
+
+# have we caught an error, and is it the one we generated?
+like( $@, qr/$warn_msg/, 'can contol');
+
+# does it indicate the right line?
+like( $@, qr/line $warn_line/, 'error location');
+}
+
+BEGIN { plan tests => $n_tests }
```
</details>

**Comment by p5pRT** [comment]
### From siegel@zrz.tu-berlin.de

Anno Siegel wrote&#8203;:
> Casey West wrote&#8203;:
> 
> > Please consider re\-submitting this patch with the proposed updates by Hugo\.
> 
> Okay\, here is a revised patch\.

\.\.\.which contained an indentation error\.  For the record\, here it is
once more\.  Sorry about that\.

Anno

<details><summary>Inline Patch</summary>

```diff
--- ../../bleadperl/warnings.pl	Thu Jun 20 18:24:36 2002
+++ warnings.pl	Sat Aug 30 20:43:41 2003
@@ -742,16 +742,17 @@
 	$i -= 2 ;
     }
     else {
-        for ($i = 2 ; $pkg = (caller($i))[0] ; ++ $i) {
-            last if $pkg ne $this_pkg ;
-        }
-        $i = 2
-            if !$pkg || $pkg eq $this_pkg ;
+        $i = _error_loc(); # see where Carp will allocate the error
     }
 
     my $callers_bitmask = (caller($i))[9] ;
     return ($callers_bitmask, $offset, $i) ;
 }
+
+sub _error_loc {
+    require Carp::Heavy;
+    goto &Carp::short_error_loc; # don't introduce another stack frame
+}                                                             
 
 sub enabled
 {
--- /dev/null	Sat Aug 30 16:32:43 2003
+++ bug_15395.t	Sat Aug 30 22:30:50 2003
@@ -0,0 +1,56 @@
+#!perl
+
+use lib 'lib';
+
+use Test::More;
+
+my $n_tests;
+BEGIN { $n_tests += 2 }
+
+my ( $warn_cat, # warning category we'll try to control
+     $warn_msg, # the error message to catch
+);
+
+{ # lexical space for tests
+    no warnings; # keep stderr clean
+
+    package SomeModule;
+    use warnings::register;
+
+    BEGIN {
+        $warn_cat = __PACKAGE__;
+        $warn_msg = 'from ' . __PACKAGE__;
+    }
+
+    # a sub that generates a random warning
+    sub gen_warning {
+        warnings::warnif( $warn_msg);
+    }
+
+    package ClientModule;
+    # use SomeModule; (would go here)
+    our @CARP_NOT = ( $warn_cat); # deliver warnings to *our* client
+
+    # call_warner provokes a warning.  it is delivered to its caller,
+    # who should also be able to control it
+    sub call_warner {
+        SomeModule::gen_warning();
+    }
+
+    # user
+
+    package main;
+    my $warn_line = __LINE__ + 3; # this line should be in the error message
+    eval {
+        use warnings FATAL => $warn_cat; # we want to know if this works
+        ClientModule::call_warner();
+    };
+
+    # have we caught an error, and is it the one we generated?
+    like( $@, qr/$warn_msg/, 'can contol');
+
+    # does it indicate the right line?
+    like( $@, qr/line $warn_line/, 'error location');
+}
+
+BEGIN { plan tests => $n_tests }
```
</details>

**Comment by p5pRT** [comment]
### From @rgs

Anno Siegel wrote&#8203;:
> Anno Siegel wrote&#8203;:
> > Casey West wrote&#8203;:
> > 
> > > Please consider re\-submitting this patch with the proposed updates by Hugo\.
> > 
> > Okay\, here is a revised patch\.
> 
> \.\.\.which contained an indentation error\.  For the record\, here it is
> once more\.  Sorry about that\.

Thanks\, and sorry for the delay\. I've applied your patch as change 21167
below\, with minimal tweaks :

I don't consider it suitable for 5\.8\.1 at this point \(and Jarkko will most
probably agree with me\.\)

Change 21167 by rgs@&#8203;rgs\-home on 2003/09/10 21&#8203;:18&#8203;:59

&nbsp;        Subject&#8203;: Re&#8203;: \[perl \#15395\] lexical warnings and inheritance
&nbsp;        From&#8203;: Anno Siegel \<Siegel@&#8203;zrz\.TU\-Berlin\.DE>
&nbsp;        Date&#8203;: Sun\, 31 Aug 2003 22&#8203;:08&#8203;:19 GMT
&nbsp;        Message\-Id&#8203;: \<200308312208\.WAA25312@&#8203;lublin\.zrz\.TU\-Berlin\.DE>

&nbsp;        The test has been moved into warnings\.pm's test suite\. Note
&nbsp;        that this patch fixed as well the behaviour of warnings&#8203;::enabled
&nbsp;        regarding lexical scoping and different files; hence the expected
&nbsp;        results in a few tests in /t/lib/warnings/9enabled has been
&nbsp;        reverted\.

Affected files \.\.\.

\.\.\. //depot/perl/lib/warnings\.pm\#33 edit
\.\.\. //depot/perl/t/lib/warnings/9enabled\#5 edit
\.\.\. //depot/perl/warnings\.pl\#37 edit

\-\-

**Comment by p5pRT** [comment]
@rgs - Status changed from 'new' to 'resolved'
